packages/react/src/file-upload/FileUpload.tsx

```diff
@@ -171,21 +194,75 @@ export const FileUpload = forwardRef<HTMLInputElement, FileUploadProps>(
 			invalid: invalid || !!errorSummary,
 		});
 
+		const valueChangeToken = JSON.stringify(
+			value.filter((file) =>
+				fileRejections.every(
+					(rejectedFile) => rejectedFile.file.path !== file.path
+				)
+			)
+		);
+
 		useEffect(() => {
-			setFileRejections(
-				dropzoneFileRejections.map(({ file, errors }) => ({
-					file,
-					errors: errors.map((error) => ({
-						code: error.code,
-						message: getFileRejectionErrorMessage(
-							error,
-							formattedMaxFileSize,
-							acceptedFilesSummary
-						),
-					})),
-				}))
-			);
-		}, [dropzoneFileRejections, formattedMaxFileSize, acceptedFilesSummary]);
+			const isOverMaxFilesLimit =
+				maxFiles && value.length + fileRejections.length > maxFiles;
+
+			if (isOverMaxFilesLimit) {
+				const tooManyFilesRejectionList: Array<RejectedFile> = value
+					.slice(maxFiles)
+					.map((file) => ({
+						file,
+						errors: [tooManyFilesError],
+					}));
+
+				const dropzoneRejections = dropzoneFileRejections.map((rej) => ({
+					...rej,
+					errors: [...rej.errors, tooManyFilesError],
```

Comment:
Could this duplicate the `tooManyFilesError` in the array? Say, if you edited the list of files after receiving the `tooManyFilesError` but left one file in that had this error. I can't see if the `errors` array is de-duped or reset anywhere but I may have missed something.

Reply:
No, because I do not pass the max files limit into the `useDropzone` utility on line 172 anymore.

So the only way that `dropzoneRejections` gets populated with values now is if the file size is too large.

Reply:
Awesome 🙌